### ISSUE TITLE
Warn about the use of non-generic collections

### ISSUE THREAD

**Issue by k4b7** [dotnet/fsharp#16259]
**Is your feature request related to a problem? Please describe.**

I'm an F# newbie and was working with some existing code that used `hashtable` which is non-generic.  I was given some advice to use `Dictionary` instead by to be careful to not open `System.Collections` because it contains a non-generic version of `Dictionary`.

**Describe the solution you'd like**

As an F# newbie it would be nice if the compiler warned about this sort of thing.

**Describe alternatives you've considered**

I could try creating an FSharpLint rule for this.

**Additional context**

N/A

**Comment by charlesroddie** [comment]
This is for FsLang/Suggestions.

It's likely that "Analyzer" is going to be the answer.

You could warn on say a non-generic list
1. When it is instantiated via a blacklist for specific types.
2. When something is added to it with an implicit cast to `obj`, since the Add method takes an `obj`.
3. When something is taken out of it and needs to be cast to a type via for example `:?>`.

3. should definitely get a warning by default, 2. should get a warning too but with an escape hatch (e.g. it's valid to collect a group of `obj`s for the purpose of using their `ToString` methods).

1. would be useful as it would give the earliest warning but it's rather ad-hoc. Ordinarily it's up to types to mark themselves as deprecated or not visible to intellisense. Effectively this would be a correction to the lack of such an annotation in dotnet/runtime. We should ask them first whether there is a plan to do this before going our own way on this.

**Comment by vzarytovskii** [comment]
Yeah, should likely be analyzer, since it's too situational and opinionated.

**Comment by T-Gro** [comment]
Yes, added it to the tracking issue for "solvable with analyzers".
